sdk/go/common/util/result/result.go

```diff
@@ -19,8 +19,43 @@ import (
 	"fmt"
 
 	"github.com/hashicorp/go-multierror"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 )
 
+type bailError struct {
+	err error
+}
+
+// BailError is the replacement for Result now that Go supports wrapping errors.  It is used to indicate that
+// a computation failed, but that it failed gracefully, i.e. it is not a bug in Pulumi. BailError implements
+// the error interface but will prefix it's error string with BAIL, which if ever seen in user facing messages
+// indicates that a check for bailing was missed. It also blocks `Unwrap` calls, instead to get access to the
+// inner error use the `IsBail` function.
+func BailError(err error) error {
+	contract.Requiref(err != nil, "err", "must not be nil")
+
+	return &bailError{err: err}
+}
+
+func (b *bailError) Error() string {
+	return fmt.Sprintf("BAIL: %v", b.err)
+}
+
+// IsBail returns the inner error if the error passed in was a BailError, else nil.
```

Comment:
I'm slightly torn on this API. Really the inner error is just for debugging, once you've returned a bail error nothing higher up the callstack should be looking at the error that trigger it. So maybe this should just be a book result?

Reply:
> I'm curious why BailError doesn't want to support unwrapping, but no big deal.

Because I want to treat the bail error as the final result, the internal error is _just_ for debugging. Once we've bailed the rest of the system shouldn't care _why_ we bailed, but it's incredibly useful to keep track of it for debugging so a human can care.

I'll change IsBail to just a bool result, and not add Unbail. We'll just have the inner error for humans to see during debugging.

I think Bailf makes sense, as with FprintBailf I'd consider it but was on the edge of if it was worth adding or not.